[Alerting] Add popover for Elasticsearch query rule type

lcawl opened the pull request:
## Summary

Relates to https://github.com/elastic/kibana/pull/134763

That PR included edits to the UI text for the Elasticsearch query rule type and removed a tooltip associated with the query threshold fields.

This PR:

- Adds a popover to provide more detailed advice about the query threshold and time window, based on information that currently exists only in https://www.elastic.co/guide/en/kibana/master/rule-type-es-query.html
- Changes the term "duration" with "time window". I recommended the use of duration in the previous PR but now think it's misleading.

### Screenshots

![image](https://user-images.githubusercontent.com/26471269/179115072-b262af9d-7125-4b6c-9f86-4a90a4588534.png)

### Checklist

Delete any items that are not applicable to this PR.

- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)

### For maintainers

- [ ] This was checked for breaking API changes and was [labeled appropriately](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)

elasticmachine commented:
Pinging @elastic/response-ops (Team:ResponseOps)

lcawl commented on x-pack/plugins/stack_alerts/public/alert_types/es_query/rule_common_expressions/threshold_help_popover.tsx:
```diff
@@ -0,0 +1,106 @@
+          <p>
+            <FormattedMessage
+              id="xpack.stackAlerts.esQuery.ui.thresholdHelp.duplicateMatches"
+              defaultMessage="This rule type checks for duplication of document matches across multiple runs. If you configure the rule with a time window greater than the check interval and a document matches the query in multiple runs, it is considered in only the first threshold calculation. The rule uses the timestamp of the matches to avoid alerting on the same match multiple times."
```
I think we can likely reduce this callout further. Perhaps the last sentence is unnecessary?
```suggestion
              defaultMessage="This rule type checks for duplication of document matches across multiple runs. If you configure the rule with a time window greater than the check interval and a document matches the query in multiple runs, it is considered in only the first threshold calculation."
```

ymao1 replied:
I like this simplification.

gchaps reviewed (commented):
This is a nice start.   Here are a few comments.

**Placement**

Can this pop-up cover just the greyed out background and not the Create rule form? It would nice to see the rest of the form for context.

We should look at other help pop-ups and use the same width.

**Formatting**

- Larger font size for title
- Spacing between the two steps


**Text**

Overall, I think we can cut down the text a bit.

mdefazio commented:
**A few thoughts:** 

I think we'll at least want to put a max-width on the popover panel to account for wide screens...

<img width="2502" alt="image" src="https://user-images.githubusercontent.com/3756330/178068614-f48168d1-c37a-43a5-8c0b-588a183b13d6.png">

What if we use a tour component instead? It has a maxW prop and we can also break apart the steps based on the input (the multi doc callout might prove tricky though).  

I'm wondering if this helps us a) avoid taking over the screen and b) is readable in the context of the input (depending on the screen size and how far I've scrolled when I click the help, it may hide the inputs). 

<img width="789" alt="image" src="https://user-images.githubusercontent.com/3756330/178071931-08ecf698-4f7e-4966-a5ed-3939944dea6c.png">

Here's a rough breakdown of what I was imagining. 
[Whimsical link](https://whimsical.com/docs-popover-BD21gVdWc6xdYgXaSLmHqE)

<img width="1949" alt="image" src="https://user-images.githubusercontent.com/3756330/178071606-d8f30ca4-a6ec-45d8-8f20-30b71520e20c.png">

Granted, I think this does beg the question of whether this should go through the whole form...

mdefazio commented on x-pack/plugins/stack_alerts/public/alert_types/es_query/rule_common_expressions/threshold_help_popover.tsx:
```diff
@@ -0,0 +1,106 @@
+
+  _renderContent() {
+    return (
+      <div>
```
To keep the popover from expanding the width of the screen, perhaps just adding a max-width will move this forward. 

<img width="700" alt="image" src="https://user-images.githubusercontent.com/3756330/178768102-be3fe03a-20f3-49fe-937c-90415bc5e7ab.png">

Can do it inline here with a style tag, or via emotion (for just one property, maybe overkill, but makes it easier to find in the code)

```
import { css } from '@emotion/react';

const PopoverStyles = css`
max-width: 400px;
`

<div css={PopoverStyles}> ... 


```


I had used 400px in the screenshot.  Arbitrary number, so feel free to adjust if you feel necessary.

gchaps replied:
@mdefazio I having a hard time with the pop-up covering the fields I asked for help on.  Can we change its placement?

lcawl replied:
It might help if we put the icon before the label (at the left side of the UI). I'll give that a try.

lcawl replied:
Thanks for the information!  I've added that css change and to avoid covering up the fields I've changed `anchorPosition` to `upLeft`.

mdefazio reviewed (changes requested):
Just have the one comment on the width of the popover. 

I think we can explore some other options on incorporating this copy within the form. But this certainly is an improvement.

lcawl commented:
> Can this pop-up cover just the greyed out background and not the Create rule form? It would nice to see the rest of the form for context.

I didn't see how to make that happen, but I did change the anchor position so it's less likely to cover the pertinent fields.
 
> * Larger font size for title

I was unable to force the popover title and the callout title to match, so I've removed the callout title.

> * Re: the steps, @KOTungseth reminded me that when we first started implementing in-product assistance, we talked about leaving step-by-step content in the product docs...

Sure, I've changed it back to paragraphs.

> Overall, I think we can cut down the text a bit.

I've reduced the callout to a single line. I think we could likely also drop the "For example..." phrase, but if you see other areas to cut, let me know!

lcawl commented on x-pack/plugins/stack_alerts/public/alert_types/es_query/rule_common_expressions/threshold_help_popover.tsx:
```diff
@@ -0,0 +1,112 @@
+          <p>
+            <FormattedMessage
+              id="xpack.stackAlerts.esQuery.ui.thresholdHelp.threshold"
+              defaultMessage="Each time the rule runs, it checks whether the number of documents that match your query meets this threshold. For example, IS ABOVE 1000."
```
```suggestion
              defaultMessage="Each time the rule runs, it checks whether the number of documents that match your query meets this threshold."
```
```suggestion
              defaultMessage="Each time the rule runs, it checks whether the number of documents that match your query meets this threshold. For example, IS ABOVE 1000."
```

mdefazio reviewed (approved):
LGTM!

gchaps commented:
Looking good.  

I think we should be more direct in the second paragraph:

> To avoid gaps in detection, set this value greater than or equal to the value you chose for the Check every field.

Hopefully, the pop-up doesn't cover the "Check every" field.

I like the blue box at the bottom and without the title.  Is there a way we can indicate this is a note?  Maybe an icon and the word note:

> x **Note**: If the time window ....

gchaps reviewed (approved):
LGTM

lcawl commented:
> I think we should be more direct in the second paragraph:

Thanks!  I made that change in c6592f63ef07a7c2ec53e77b20378652ce4df52b

> Hopefully, the pop-up doesn't cover the "Check every" field.

Unfortunately the "check every" field is right near the top of that UI so folks have likely scrolled too far to see it anymore. Hopefully it's still fresh in their memory.

> I like the blue box at the bottom and without the title. Is there a way we can indicate this is a note? Maybe an icon and the word note:

Since it's a single sentence in that callout now, I was able to get the icon to re-appear (it disappeared when I removed the title) by turning the sentence into the title. I've refreshed the initial screenshot with the result.
